pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,84 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	logger "github.com/rs/zerolog/log"
+)
+
+var (
+	c           *collector
+	metricStore MetricStore
+	log         zerolog.Logger
+
+	mu   = &sync.Mutex{}
+	once sync.Once
+)
+
+func init() {
+	log = logger.With().
+		Str("component", "telemetry").
+		Logger()
+}
+
+// MetricStore specifies the methods for persisting a metric.
+type MetricStore interface {
+	StoreMetric(context.Context, Metric) error
+	Close() error
+}
+
+// SetMetricStore sets the store implementation.
+// Can only be called once, and should be called before Collect.
+func SetMetricStore(s MetricStore) {
+	once.Do(func() {
+		metricStore = s
+	})
+}
+
+// Collect collects the metric by persisting locally for later publication.
+// If Collect is called before setting the metric store, it will simply log the metric without persisting it.
+func Collect(ctx context.Context, metric interface{}) error {
```

Comment:
I opted for a global function on this library to make it really easy to collect a metric in a non-invasive manner. You can simply call `telemetry.Collect`. No need to inject any dependency on clients. Of course that comes with some trade-off. We have the [global variables](https://github.com/tablelandnetwork/go-tableland/pull/264/files#diff-1daf3b94af9e46763e64ff02c31240f3c3b1421994d4d56e05769dfd4f765ba4R13-R20) that mess with concurrency, and make the code error prone.